src/core/execute.c

```diff
@@ -7040,33 +7040,56 @@ void exec_directory_done(ExecDirectory *d) {
         d->mode = 0755;
 }
 
-int exec_directory_add(ExecDirectoryItem **d, size_t *n, const char *path, char **symlinks) {
+static ExecDirectoryItem *exec_directory_find(ExecDirectory *d, const char *path) {
+        assert(d);
+        assert(path);
+
+        for (size_t i = 0; i < d->n_items; i++)
+                if (path_equal(d->items[i].path, path))
+                        return &d->items[i];
+
+        return NULL;
+}
+
+int exec_directory_add(ExecDirectory *d, const char *path, const char *symlink) {
         _cleanup_strv_free_ char **s = NULL;
         _cleanup_free_ char *p = NULL;
+        ExecDirectoryItem *existing;
+        int r;
 
         assert(d);
-        assert(n);
         assert(path);
 
+        existing = exec_directory_find(d, path);
+        if (existing) {
+                if (symlink) {
+                        r = strv_extend(&existing->symlinks, symlink);
+                        if (r < 0)
+                                return r;
+                }
```

Comment:
you can drop this conditionalization. strv_extend() is fine with the second param being NULL, and then is a NOP